### ISSUE TITLE
[no-unnecessary-type-assertion] doesn't handle types properly -> React['key']

### ISSUE THREAD

**Issue by Hotell** [typescript-eslint/typescript-eslint#982]
**Repro**

`no-unnecessary-type-assertion` is resolving `key` keys on React components  inconsistently with normal function call. Please take a look at the code for better descrition.

 

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

```ts
const appTest = () => {
  type ArtificialKey = string | number;
  interface ReactAttributeArtificial {
    key?: ArtificialKey;
  }
  type Data = {
    id?: null | string | number;
  };

  const Test = (props: Data) => {
    // ✅no tsc errors
    // ✅no eslint errors
    return ChildFn({ key: props.id! });
  };

  const Test2 = (props: Data) => {
    // 🚨ESLINT error.
    // 🚨If you remove non null operator it will be an compile error
    return <ChildCmp key={props.id!} />;
  };

  const ChildFn = (props: ReactAttributeArtificial) => 'Hello';
  const ChildCmp = () => <div>Hello</div>;
};
```

**Expected Result**

It should behave consistently

**Actual Result**

JSX `key` is resolved in not correct way ?

![image](https://user-images.githubusercontent.com/1223799/65064080-ed7d5a00-d97f-11e9-9a4e-12e16ebaa3db.png)

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.2.0` |
| `@typescript-eslint/parser`        | `2.2.0` |
| `TypeScript`                       | `3.6.3` |
| `ESLint`                           | `6.3.0` |
| `node`                             | `10` |
| `yarn`                              | `1.17` |

**Comment by bradzacher** [comment]
If I had to hazard a guess, I'd say that it'd probably be something to do with the rule not correctly separating `null` from `undefined`.

I'll look into it.

**Comment by bradzacher** [comment]
Hmmm..
Looking into this against master - I cannot repro this:
![image](https://user-images.githubusercontent.com/7462525/65264593-ce87e080-dac3-11e9-8fca-b2787f23a3f1.png)

Could you please share your tsconfig with me?

**Comment by Hotell** [comment]
hey @bradzacher sorry for delayed answer! 

here it is:

```json
{
  "compilerOptions": {
    "lib": ["dom", "es2017", "dom.iterable"],
    "allowJs": true,
    "noEmit": true,
    "strict": true,
    "noUnusedLocals": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "target": "esnext",
    "jsx": "preserve",
    "module": "esnext",
    "moduleResolution": "node",
    "forceConsistentCasingInFileNames": true,
    "sourceMap": true,
    "baseUrl": ".",
    "paths": {
      "assets/*": ["src/assets/*"],
      "*": ["src/js/*", "*"]
    }
  },
  "include": [
    "typings/**/*.d.ts",
    "src/js/**/*.ts",
    "src/js/**/*.tsx",
    "src/js/**/*.js",
    "apps/**/*.ts",
    "apps/**/*.tsx",
    "libs"
  ],
  "exclude": ["node_modules"]
}
```
